text/3512-closure-move-bindings.md

```diff
@@ -0,0 +1,284 @@
+- Feature `closure_move_bindings`
+- Start Date: 2023-10-09
+- RFC PR: [rust-lang/rfcs#3512](https://github.com/rust-lang/rfcs/pull/3512)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Adds the syntax `move(bindings) |...| ...`
+to explicitly specify how to capture bindings into a closure.
+
+# Motivation
+[motivation]: #motivation
+
+Currently there are two ways to capture local bindings into a closure,
+namely by reference (`|| foo`) and by moving (`move || foo`).
+This mechanism has several ergonomic problems:
+
+- It is not possible to move some bindings and reference the others.
+To do so, one must define another binding that borrows the value
+and move it into the closure:
+
+```rs
+{
+    let foo = &foo;
+    move || run(foo, bar)
+}
+```
+
+- It is a very frequent scenario to clone a value into a closure
+(especially common with `Rc`/`Arc`-based values),
+but even the simplest scenario requires three lines of boilerplate:
+
+```rs
+{
+    let foo = foo.clone();
+    move || foo.run()
+}
+```
+
+This RFC proposes a more concise syntax to express these moving semantics.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A closure may capture bindings in its defining scope.
+Bindings are captured by reference by default:
+
+```rs
+let mut foo = 1;
+let mut closure = || { foo = 2; };
+closure();
+dbg!(foo); // foo is now 2
+```
+
+You can add a `move` keyword in front of the closure
+to indicate that all captured bindings are moved into the closure
+instead of referenced:
+
+```rs
+let mut foo = 1;
+let mut closure = move || { foo = 2; };
+closure();
+dbg!(foo); // foo is still 1, but the copy of `foo` in `closure` is 2
+```
+
+Note that `foo` is _copied_ during move in this example
+as `i32` implements `Copy`.
+
+If a closure captures multiple bindings,
+the `move` keyword makes them all captured by moving.
+To move only specific bindings,
+list them in parentheses after `move`:
+
+```rs
+let foo = 1;
+let mut bar = 2;
+let mut closure = move(mut foo) || {
```

Comment:
I just noticed that GitHub renders `move()` with the color of a function call instead of as a keyword. This adds to the previously discussed problem of ambiguity as a function call.

Reply:
`move` is already a keyword and cannot be the identifier of a function, so `move(` unambiguously starts a captures list.